File: sdk/typescript/src/encryption.ts
/**
 * AES-256-GCM encryption utilities for Helix
 * Uses Web Crypto API for browser-native encryption
 */

/**
 * Encryption key wrapper
 */
export type EncryptionKey = CryptoKey;

/**
 * Encrypted data with IV
 */
export interface EncryptedData {
    /** Combined IV + ciphertext + auth tag */
    data: ArrayBuffer;
    /** IV used for encryption (first 12 bytes of data) */
    iv: Uint8Array;
}

/** IV length in bytes (96 bits) */
const IV_LENGTH = 12;

/** Algorithm configuration */
const ALGORITHM = {
    name: 'AES-GCM',
    length: 256,
};

/**
 * Encryption utilities for AES-256-GCM
 */
export class HelixEncryption {
    /**
     * Generate a new random encryption key
     * @returns CryptoKey for AES-256-GCM
     */
    async generateKey(): Promise<EncryptionKey> {
        return crypto.subtle.generateKey(
            { name: ALGORITHM.name, length: ALGORITHM.length },
            true,
            ['encrypt', 'decrypt']
        );
    }

    /**
     * Encrypt data with AES-256-GCM
     * 
     * @param data - Data to encrypt
     * @param key - Encryption key
     * @returns Encrypted data with IV prepended
     * 
     * @example
     * ```typescript
     * const key = await encryption.generateKey();
     * const encrypted = await encryption.encrypt(fileData, key);
     * // Format: [IV (12 bytes)][Ciphertext][Auth Tag (16 bytes)]
     * ```
     */
    async encrypt(data: ArrayBuffer, key: EncryptionKey): Promise<EncryptedData> {
        const iv = crypto.getRandomValues(new Uint8Array(IV_LENGTH));

        const ciphertext = await crypto.subtle.encrypt(
            { name: ALGORITHM.name, iv },
            key,
            data
        );

        const combined = new Uint8Array(iv.length + ciphertext.byteLength);
        combined.set(iv);
        combined.set(new Uint8Array(ciphertext), iv.length);

        return {
            data: combined.buffer,
            iv,
        };
    }

    /**
     * Decrypt data encrypted with AES-256-GCM
     * 
     * @param encryptedData - Data with IV prepended
     * @param key - Decryption key
     * @returns Decrypted data
     */
    async decrypt(encryptedData: ArrayBuffer, key: EncryptionKey): Promise<ArrayBuffer> {
        const dataArray = new Uint8Array(encryptedData);

        if (dataArray.length < IV_LENGTH + 16) {
            throw new Error('Invalid encrypted data: too short');
        }

        const iv = dataArray.slice(0, IV_LENGTH);
        const ciphertext = dataArray.slice(IV_LENGTH);

        return crypto.subtle.decrypt(
            { name: ALGORITHM.name, iv },
            key,
            ciphertext
        );
    }

    /**
     * Export key to base64 string for storage
     * @param key - CryptoKey to export
     * @returns Base64 encoded key
     */
    async exportKey(key: EncryptionKey): Promise<string> {
        const raw = await crypto.subtle.exportKey('raw', key);
        return Buffer.from(raw).toString('base64');
    }

    /**
     * Import key from base64 string
     * @param keyString - Base64 encoded key
     * @returns CryptoKey for decryption
     */
    async importKey(keyString: string): Promise<EncryptionKey> {
        const raw = Buffer.from(keyString, 'base64');

        if (raw.length !== 32) {
            throw new Error('Invalid key length: expected 32 bytes');
        }

        return crypto.subtle.importKey(
            'raw',
            raw,
            { name: ALGORITHM.name },
            true,
            ['encrypt', 'decrypt']
        );
    }

    /**
     * Generate a key derivation from password (for backup)
     * 
     * @param password - User password
     * @param salt - Salt for derivation
     * @returns Derived key
     */
    async deriveKey(password: string, salt: Uint8Array): Promise<EncryptionKey> {
        const passwordKey = await crypto.subtle.importKey(
            'raw',
            new TextEncoder().encode(password),
            'PBKDF2',
            false,
            ['deriveBits', 'deriveKey']
        );

        return crypto.subtle.deriveKey(
            {
                name: 'PBKDF2',
                salt,
                iterations: 100000,
                hash: 'SHA-256',
            },
            passwordKey,
            { name: ALGORITHM.name, length: ALGORITHM.length },
            true,
            ['encrypt', 'decrypt']
        );
    }

    /**
     * Generate a random salt for key derivation
     * @returns 16-byte salt
     */
    generateSalt(): Uint8Array {
        return crypto.getRandomValues(new Uint8Array(16));
    }

    /**
     * Encrypt a string (helper for filenames, etc.)
     * @param text - String to encrypt
     * @param key - Encryption key
     * @returns Base64 encoded encrypted data
     */
    async encryptString(text: string, key: EncryptionKey): Promise<string> {
        const data = new TextEncoder().encode(text);
        const encrypted = await this.encrypt(data.buffer, key);
        return Buffer.from(encrypted.data).toString('base64');
    }

    /**
     * Decrypt a string
     * @param encryptedBase64 - Base64 encoded encrypted data
     * @param key - Decryption key
     * @returns Decrypted string
     */
    async decryptString(encryptedBase64: string, key: EncryptionKey): Promise<string> {
        const data = Buffer.from(encryptedBase64, 'base64');
        const decrypted = await this.decrypt(data.buffer, key);
        return new TextDecoder().decode(decrypted);
    }
}

/**
 * Export a key to base64 string (standalone function)
 */
export async function exportKey(key: EncryptionKey): Promise<string> {
    const raw = await crypto.subtle.exportKey('raw', key);
    return Buffer.from(raw).toString('base64');
}

/**
 * Import a key from base64 string (standalone function)
 */
export async function importKey(keyString: string): Promise<EncryptionKey> {
    const raw = Buffer.from(keyString, 'base64');
    return crypto.subtle.importKey(
        'raw',
        raw,
        { name: ALGORITHM.name },
        true,
        ['encrypt', 'decrypt']
    );
}

/**
 * Encrypt data with a base64 key string
 */
export async function encryptWithKeyString(
    data: ArrayBuffer,
    keyString: string
): Promise<ArrayBuffer> {
    const key = await importKey(keyString);
    const iv = crypto.getRandomValues(new Uint8Array(IV_LENGTH));
    const ciphertext = await crypto.subtle.encrypt(
        { name: ALGORITHM.name, iv },
        key,
        data
    );
    const combined = new Uint8Array(iv.length + ciphertext.byteLength);
    combined.set(iv);
    combined.set(new Uint8Array(ciphertext), iv.length);
    return combined.buffer;
}

/**
 * Decrypt data with a base64 key string
 */
export async function decryptWithKeyString(
    encryptedData: ArrayBuffer,
    keyString: string
): Promise<ArrayBuffer> {
    const key = await importKey(keyString);
    const dataArray = new Uint8Array(encryptedData);
    const iv = dataArray.slice(0, IV_LENGTH);
    const ciphertext = dataArray.slice(IV_LENGTH);
    return crypto.subtle.decrypt(
        { name: ALGORITHM.name, iv },
        key,
        ciphertext
    );
}

// AES-256-GCM

File: sdk-python/helix_sdk/encryption.py
"""
AES-256-GCM encryption utilities for Helix

Provides symmetric encryption for file content and metadata,
compatible with the TypeScript SDK.
"""

from __future__ import annotations

import os
import base64
from typing import Tuple

from cryptography.hazmat.primitives.ciphers.aead import AESGCM


IV_LENGTH = 12
KEY_LENGTH = 32
TAG_LENGTH = 16


class HelixEncryption:
    """
    Encryption utilities for AES-256-GCM.
    
    Compatible with Web Crypto API used in the TypeScript SDK.
    
    Example:
        >>> enc = HelixEncryption()
        >>> key = enc.generate_key()
        >>> encrypted = enc.encrypt(b"secret data", key)
        >>> decrypted = enc.decrypt(encrypted, key)
    """
    
    def generate_key(self) -> bytes:
        """
        Generate a new random 256-bit encryption key.
        
        Returns:
            32-byte random key
        """
        return os.urandom(KEY_LENGTH)
    
    def encrypt(self, data: bytes, key: bytes) -> bytes:
        """
        Encrypt data with AES-256-GCM.
        
        The IV is prepended to the ciphertext, matching the format
        used by the TypeScript SDK.
        
        Args:
            data: Data to encrypt
            key: 32-byte encryption key
            
        Returns:
            IV + ciphertext + auth tag
        """
        if len(key) != KEY_LENGTH:
            raise ValueError(f"Key must be {KEY_LENGTH} bytes")
        
        iv = os.urandom(IV_LENGTH)
        aesgcm = AESGCM(key)
        ciphertext = aesgcm.encrypt(iv, data, None)
        
        return iv + ciphertext
    
    def decrypt(self, encrypted_data: bytes, key: bytes) -> bytes:
        """
        Decrypt data encrypted with AES-256-GCM.
        
        Expects the IV to be prepended to the ciphertext.
        
        Args:
            encrypted_data: IV + ciphertext + auth tag
            key: 32-byte decryption key
            
        Returns:
            Decrypted data
        """
        if len(key) != KEY_LENGTH:
            raise ValueError(f"Key must be {KEY_LENGTH} bytes")
        
        if len(encrypted_data) < IV_LENGTH + TAG_LENGTH:
            raise ValueError("Encrypted data too short")
        
        iv = encrypted_data[:IV_LENGTH]
        ciphertext = encrypted_data[IV_LENGTH:]
        
        aesgcm = AESGCM(key)
        return aesgcm.decrypt(iv, ciphertext, None)
    
    def export_key(self, key: bytes) -> str:
        """
        Export a key to base64 string for storage.
        
        Args:
            key: Raw key bytes
            
        Returns:
            Base64 encoded key
        """
        return base64.b64encode(key).decode("utf-8")
    
    def import_key(self, key_string: str) -> bytes:
        """
        Import a key from base64 string.
        
        Args:
            key_string: Base64 encoded key
            
        Returns:
            Raw key bytes
        """
        key = base64.b64decode(key_string)
        if len(key) != KEY_LENGTH:
            raise ValueError(f"Invalid key length: expected {KEY_LENGTH} bytes")
        return key
    
    def encrypt_string(self, text: str, key: bytes) -> str:
        """
        Encrypt a string and return base64 encoded result.
        
        Args:
            text: String to encrypt
            key: Encryption key
            
        Returns:
            Base64 encoded encrypted data
        """
        data = text.encode("utf-8")
        encrypted = self.encrypt(data, key)
        return base64.b64encode(encrypted).decode("utf-8")
    
    def decrypt_string(self, encrypted_b64: str, key: bytes) -> str:
        """
        Decrypt a base64 encoded string.
        
        Args:
            encrypted_b64: Base64 encoded encrypted data
            key: Decryption key
            
        Returns:
            Decrypted string
        """
        encrypted = base64.b64decode(encrypted_b64)
        decrypted = self.decrypt(encrypted, key)
        return decrypted.decode("utf-8")


def generate_key() -> bytes:
    """Generate a new random 256-bit encryption key."""
    return os.urandom(KEY_LENGTH)


def encrypt_data(data: bytes, key: bytes) -> bytes:
    """
    Encrypt data with AES-256-GCM.
    
    Args:
        data: Data to encrypt
        key: 32-byte encryption key
        
    Returns:
        IV + ciphertext + auth tag
    """
    if len(key) != KEY_LENGTH:
        raise ValueError(f"Key must be {KEY_LENGTH} bytes")
    
    iv = os.urandom(IV_LENGTH)
    aesgcm = AESGCM(key)
    ciphertext = aesgcm.encrypt(iv, data, None)
    
    return iv + ciphertext


def decrypt_data(encrypted_data: bytes, key: bytes) -> bytes:
    """
    Decrypt data encrypted with AES-256-GCM.
    
    Args:
        encrypted_data: IV + ciphertext + auth tag
        key: 32-byte decryption key
        
    Returns:
        Decrypted data
    """
    if len(key) != KEY_LENGTH:
        raise ValueError(f"Key must be {KEY_LENGTH} bytes")
    
    if len(encrypted_data) < IV_LENGTH + TAG_LENGTH:
        raise ValueError("Encrypted data too short")
    
    iv = encrypted_data[:IV_LENGTH]
    ciphertext = encrypted_data[IV_LENGTH:]
    
    aesgcm = AESGCM(key)
    return aesgcm.decrypt(iv, ciphertext, None)


def export_key(key: bytes) -> str:
    """Export a key to base64 string."""
    return base64.b64encode(key).decode("utf-8")


def import_key(key_string: str) -> bytes:
    """Import a key from base64 string."""
    key = base64.b64decode(key_string)
    if len(key) != KEY_LENGTH:
        raise ValueError(f"Invalid key length: expected {KEY_LENGTH} bytes")
    return key


def derive_key_from_password(
    password: str,
    salt: bytes,
    iterations: int = 100000,
) -> bytes:
    """
    Derive an encryption key from a password using PBKDF2.
    
    Args:
        password: User password
        salt: Random salt (should be stored with encrypted data)
        iterations: Number of PBKDF2 iterations
        
    Returns:
        32-byte derived key
    """
    from cryptography.hazmat.primitives import hashes
    from cryptography.hazmat.primitives.kdf.pbkdf2 import PBKDF2HMAC
    
    kdf = PBKDF2HMAC(
        algorithm=hashes.SHA256(),
        length=KEY_LENGTH,
        salt=salt,
        iterations=iterations,
    )
    
    return kdf.derive(password.encode("utf-8"))


def generate_salt(length: int = 16) -> bytes:
    """Generate a random salt for key derivation."""
    return os.urandom(length)


class KeyStorage:
    """
    Simple key storage for managing encryption keys.
    
    Keys are stored in memory with optional file backup.
    """
    
    def __init__(self):
        self._keys: dict[str, bytes] = {}
    
    def store(self, transaction_id: str, key: bytes) -> None:
        """Store a key for a transaction."""
        self._keys[transaction_id] = key
    
    def get(self, transaction_id: str) -> bytes | None:
        """Get a stored key by transaction ID."""
        return self._keys.get(transaction_id)
    
    def delete(self, transaction_id: str) -> bool:
        """Delete a stored key."""
        if transaction_id in self._keys:
            del self._keys[transaction_id]
            return True
        return False
    
    def export_all(self) -> dict[str, str]:
        """Export all keys as base64 strings."""
        return {
            tx_id: export_key(key)
            for tx_id, key in self._keys.items()
        }
    
    def import_all(self, keys: dict[str, str]) -> None:
        """Import keys from base64 strings."""
        for tx_id, key_b64 in keys.items():
            self._keys[tx_id] = import_key(key_b64)
    
    def save_to_file(self, path: str) -> None:
        """Save keys to a JSON file."""
        import json
        from pathlib import Path
        
        data = self.export_all()
        Path(path).write_text(json.dumps(data, indent=2))
    
    def load_from_file(self, path: str) -> None:
        """Load keys from a JSON file."""
        import json
        from pathlib import Path
        
        data = json.loads(Path(path).read_text())
        self.import_all(data)

# Encryption utilities
